lib/postgres_v3_experimental.dart

```diff
@@ -0,0 +1,267 @@
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:collection/collection.dart';
+import 'package:stream_channel/stream_channel.dart';
+
+import 'src/v3/connection.dart';
+import 'src/v3/protocol.dart';
+import 'src/v3/query_description.dart';
+import 'src/v3/types.dart';
+
+export 'src/v3/types.dart';
+
+abstract class PgPool implements PgSession, PgSessionExecutor {
+  static Future<PgPool> open(
+    List<PgEndpoint> endpoints, {
+    PgSessionSettings? sessionSettings,
+    PgPoolSettings? poolSettings,
+  }) =>
+      throw UnimplementedError();
+
+  Future<R> withConnection<R>(
+    Future<R> Function(PgConnection connection) fn, {
+    PgSessionSettings? sessionSettings,
+  });
+}
+
+class PgQueryDescription {
+  factory PgQueryDescription.direct(String sql, {List<PgDataType>? types}) =
+      InternalQueryDescription.direct;
+  factory PgQueryDescription.map(String sql, {String substitution}) =
+      InternalQueryDescription.map;
+}
+
+abstract class PgSession {
+  // uses extended query protocol
+  Future<PgStatement> prepare(
+    Object /* String | InternalQueryDescription */ query, {
+    Duration? timeout,
+  });
+
+  Future<PgResult> execute(
+    Object /* String | InternalQueryDescription */ query, {
+    Object? /* List<Object?|PgTypedParameter> | Map<String, Object?|PgTypedParameter> */ parameters,
+    Duration? timeout,
+  });
+
+  Future<void> close();
+}
+
+abstract class PgSessionExecutor {
+  // TODO: also add retry options similarly to postgres_pool
+  Future<R> run<R>(Future<R> Function(PgSession session) fn);
+  Future<R> runTx<R>(Future<R> Function(PgSession session) fn);
+}
+
+abstract class PgConnection implements PgSession, PgSessionExecutor {
+  static Future<PgConnection> open(
+    PgEndpoint endpoint, {
+    PgSessionSettings? sessionSettings,
+  }) {
+    return PgConnectionImplementation.connect(endpoint,
+        sessionSettings: sessionSettings);
+  }
+
+  PgChannels get channels;
```

Comment:
I have also removed the `messages` getter from the original draft API.

Since we're now using `StreamChannel`, we can let the user specify a `StreamChannelTransformer` instead. This is much more powerful, since it can be used to implement any logic on the connection. Messages can not only be read, but also altered, dropped or injected.